docs/configure.md
# Config

When deploying the service, set configuration variables by including a `config.json` in the
root. Available configuration options:

- `timeout` _default `10000`_ - set the timeout used to render the target page.
- `port` _default `3000`_ - set the port to use for running and listening the rendertron service. Note if process.env.PORT is set, it will be used instead.
- `host` _default `0.0.0.0`_ - set the hostname to use for running and listening the rendertron service. Note if process.env.HOST is set, it will be used instead.
- `width` _default `1000`_ - set the width (resolution) to be used for rendering the page.
- `height` _default `1000`_ - set the height (resolution) to be used for rendering the page.
- `cache` _default `null`_ - set to `datastore` to enable caching on Google Cloud using datastore _only use if deploying to google cloud_, `memory` to enable in-memory caching or `filesystem` to enable disk based caching
- `cacheConfig` - an object array to specify caching options
- `renderOnly` - restrict the endpoint to only service requests for certain domains. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
- `renderOnlyPattern` - similar to `renderOnly` but uses a regex instead of a strict prefix match. eg. `"^https://.*mysite.com.*"`
- `closeBrowser`_default `false`_ - `true` forces the browser to close and reopen between each page render, some sites might need this to prevent URLs past the first one rendered returning null responses.

## cacheConfig

- `cacheDurationMinutes` _default `1440`_ - set an expiry time in minues, defaults to 24 hours. Set to -1 to disable cache Expiration
- `cacheMaxEntries` _default `100`_ - set the maximum number of entries stored in the selected cache method. Set to `-1` to allow unlimited caching. If using the datastore caching method, setting this value over `1000` may lead to degraded performance as the query to determine the size of the cache may be too slow. If you want to allow a larger cache in `datastore` consider setting this to `-1` and managing the the size of your datastore using a method like this [Deleting Entries in Bulk](https://cloud.google.com/datastore/docs/bulk-delete)
- `snapshotDir` _default `<your os's default tmp dir>/renderton`_ - **filesystem only** the directory the rendertron caches will be stored in

### Example

An example config file specifying a memory cache, with a 2 hour expiration, and a maximum of 50 entries

```javascript
{
    "cache": "memory",
    "cacheConfig": {
        "cacheDurationMinutes": 120,
        "cacheMaxEntries": 50
    }
}
```


CHANGELOG.md
# Change Log

<!-- ## Unreleased -->

## [3.1.0] 2020-11-04

### Security improvements

- Prevents rendering of \*.internal URLs, patching an issue with GCE.
- Adds allow-listing option to restrict rendering to a given list of domains or URL patterns.

### Features

- Include Heroku deploy documentation and an interactive "Click to deploy" for Heroku
- Introduces an API endpoint to clear all cache
- Adds timezone support
- Adds optional forced browser restart between renders
- Adds documentation to deploy Rendertron via Docker
- Adds option to add request headers to the rendered pages in Rendertron

### Improvements

- Fixes a bug in the filesystem cache
- Fixes issue with the injected base tag
- Updates all dependencies
- Increased the required node.js version to Node.js 10+
- Updates to the FaQ

## [3.0.0] 2020-07-02

- Introduces new configuration file format for the `config.json` options (see [README.md](./README.md))
- Introduces new cache providers:
  - In-memory cache
  - File system cache
- Introduces API endpoint to invalidate cache for a URL
- Introduces a number of new configuration options
- Introduces `refreshCache` parameter to force cache update for a URL
- Relaunches browser when the browser disconnects from Puppeteer
- Now includes doctype in rendered output
- Harmonises the configuration options for caches
- Closes page after screenshot
- Fixes security issue with AppEngine deployments
- Fixes issue with specifying host and port

## [2.0.1] 2018-09-18

- Remove testing and other files from NPM package.
- Fix NPM main config.
- Improved restrictions for endpoints.
- Support for structured data by not stripping all script tags.

## [2.0.0] 2018-07-26

- Rebuilt with Puppeteer under the hood
- Rebuilt as Koa server instead of an Express server
- Rebuilt using Typescript
- Removed explicit rendering flag
- Added support for a mobile parameter
- Added more options for screenshots

## [1.1.1] 2018-01-05

- Update `debug` flag to log requested URLs to render
- Fix for renderComplete flag
- Minor bug fixes

## [1.1.0] 2017-10-27

- Initial release on NPM


src/config.ts
/*
 * Copyright 2018 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 not
 * use this file except in compliance with the License. You may obtain a copy
 of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 under
 * the License.
 */

'use strict';

import * as fse from 'fs-extra';
import * as path from 'path';
import * as os from 'os';

const CONFIG_PATH = path.resolve(__dirname, '../config.json');

export type Config = {
  cache: 'datastore' | 'memory' | 'filesystem' | null;
  cacheConfig: { [key: string]: string };
  timeout: number;
  port: string;
  host: string;
  width: number;
  height: number;
  reqHeaders: { [key: string]: string };
  headers: { [key: string]: string };
  puppeteerArgs: Array<string>;
  renderOnly: Array<string>;
  renderOnlyPattern: string | null;
  closeBrowser: boolean;
  restrictedUrlPattern: string | null;
};

export class ConfigManager {
  public static config: Config = {
    cache: null,
    cacheConfig: {
      snapshotDir: path.join(os.tmpdir(), 'rendertron'),
      cacheDurationMinutes: (60 * 24).toString(),
      cacheMaxEntries: '100',
    },
    timeout: 10000,
    port: '3000',
    host: '0.0.0.0',
    width: 1000,
    height: 1000,
    reqHeaders: {},
    headers: {},
    puppeteerArgs: ['--no-sandbox'],
    renderOnly: [],
    renderOnlyPattern: null,
    closeBrowser: false,
    restrictedUrlPattern: null
  };

  static async getConfiguration(): Promise<Config> {
    // Load config.json if it exists.
    if (fse.pathExistsSync(CONFIG_PATH)) {
      const configJson = await fse.readJson(CONFIG_PATH);

      // merge cacheConfig
      const cacheConfig = Object.assign(
        ConfigManager.config.cacheConfig,
        configJson.cacheConfig
      );

      ConfigManager.config = Object.assign(ConfigManager.config, configJson);

      ConfigManager.config.cacheConfig = cacheConfig;
    }
    return ConfigManager.config;
  }
}
